testutils/sims/runner.go

```diff
@@ -0,0 +1,233 @@
+	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation/client/cli"
+)
+
+const SimAppChainID = "simulation-app"
+
+// this list of seeds was imported from the original simulation runner: https://github.com/cosmos/tools/blob/v1.0.0/cmd/runsim/main.go#L32
+var defaultSeeds = []int64{
+	1, 2, 4, 7,
+	32, 123, 124, 582, 1893, 2989,
+	3012, 4728, 37827, 981928, 87821, 891823782,
+	989182, 89182391, 11, 22, 44, 77, 99, 2020,
+	3232, 123123, 124124, 582582, 18931893,
+	29892989, 30123012, 47284728, 7601778, 8090485,
+	977367484, 491163361, 424254581, 673398983,
+}
+
+type SimStateFactory struct {
+	Codec       codec.Codec
+	AppStateFn  simtypes.AppStateFn
+	BlockedAddr map[string]bool
+}
+
+// SimulationApp abstract app that is used by sims
+type SimulationApp interface {
+	runtime.AppSimI
+	SetNotSigverifyTx()
+	GetBaseApp() *baseapp.BaseApp
+	TxConfig() client.TxConfig
+}
+
+// Run is a helper function that runs a simulation test with the given parameters.
+// It calls the RunWithSeeds function with the default seeds and parameters.
+//
+// This is the entrypoint to run simulation tests that used to run with the runsim binary.
+func Run[T SimulationApp](
+	t *testing.T,
+	appFactory func(
+		logger log.Logger,
+		db dbm.DB,
+		traceStore io.Writer,
+		loadLatest bool,
+		appOpts servertypes.AppOptions,
+		baseAppOptions ...func(*baseapp.BaseApp),
+	) T,
+	setupStateFactory func(app T) SimStateFactory,
+	postRunActions ...func(t *testing.T, app TestInstance[T]),
+) {
+	t.Helper()
+	RunWithSeeds(t, appFactory, setupStateFactory, defaultSeeds, nil, postRunActions...)
+}
+
+// RunWithSeeds is a helper function that runs a simulation test with the given parameters.
+// It iterates over the provided seeds and runs the simulation test for each seed in parallel.
+//
+// It sets up the environment, creates an instance of the simulation app,
+// calls the simulation.SimulateFromSeed function to run the simulation, and performs post-run actions for each seed.
+// The execution is deterministic and can be used for fuzz tests as well.
+//
+// The system under test is isolated for each run but unlike the old runsim command, there is no Process separation.
+// This means, global caches may be reused for example. This implementation build upon the vanialla Go stdlib test framework.
+func RunWithSeeds[T SimulationApp](
+	t *testing.T,
+	appFactory func(
+		logger log.Logger,
+		db dbm.DB,
+		traceStore io.Writer,
+		loadLatest bool,
+		appOpts servertypes.AppOptions,
+		baseAppOptions ...func(*baseapp.BaseApp),
+	) T,
+	setupStateFactory func(app T) SimStateFactory,
+	seeds []int64,
+	fuzzSeed []byte,
+	postRunActions ...func(t *testing.T, app TestInstance[T]),
+) {
+	t.Helper()
+	cfg := cli.NewConfigFromFlags()
+	cfg.ChainID = SimAppChainID
+	for i := range seeds {
+		seed := seeds[i]
+		t.Run(fmt.Sprintf("seed: %d", seed), func(t *testing.T) {
+			t.Parallel()
```

Comment:
In fuzzing mode, this will not be useful at all. I'm not an expert in the implementation details, but it could be the case that executing parallel() will slow down the execution (e.g. during the setup of the parallel tests). It could be skipped using:
```go
if (len(seeds) > 1) {
        t.Parallel()
}
```

Reply:
Good 👁️ . I was testing with this setting without diving into the Go stdlib. But I did now and this flag is ignored during Fuzzing. See https://github.com/golang/go/blob/go1.22.4/src/testing/fuzz.go#L695 and above.

Reply:
Fair enought, let's close this comment.